jobs/integration/conftest.py

```diff
@@ -141,22 +154,52 @@ async def run(self, cmd, *args, stdin=None):
         if hasattr(stdin, "encode"):
             stdin = stdin.encode("utf8")
 
-        stdout, stderr = await proc.communicate(input=stdin)
-        if proc.returncode != 0:
+        stdout, stderr = bytearray(), bytearray()
+
+        def tee(line: bytes, sink: bytearray, fd: int):
+            sink += line
+            write |= _tee == "out" and fd == 1
+            write |= _tee == "err" and fd == 2
+            if write or _tee == True:
+                os.write(fd, line)
+
+        async def _read_stream(stream, callback):
+            while True:
+                line = await stream.readline()
+                if line:
+                    callback(line)
+                else:
+                    break
+
+        await asyncio.wait(
+            [
+                _read_stream(proc.stdout, lambda l: tee(l, stdout, 1)),
+                _read_stream(proc.stderr, lambda l: tee(l, stderr, 2)),
+            ]
+        )
```

Comment:
while reading from the process' streams

Reply:
It looks like we're waiting here for the stdout and stderr streams to end before moving on. But, we haven't sent stdin to the process. Actually, it looks like we never do that?

Indeed, if I make this a standalone function and run it like so:
```
out, err = await run("cat", stdin="foo")
```

the call hangs forever, because `cat` is waiting for stdin that never gets sent to it.

I think this would need a `_write_stream` or something, too

Reply:
i tested with this:

```python
async def test_tools_run(tools):
    out, err = await tools.run("cat", stdin="foo", _tee=True)
    assert out == "foo"
    assert err == ""
  ```